thrift.js

```diff
@@ -121,33 +133,120 @@ function Thrift(options) {
     this.dirname = path.dirname(this.filename);
     this.memo[this.filename] = this;
 
-    var ast = this.asts[options.entryPoint];
-    var source = this.idls[options.entryPoint];
-    if (!source && !ast) {
-        /* eslint-disable max-len */
-        assert.ok(this.fs, this.filename + ': Thrift must be constructed with either a complete set of options.idls, options.asts, or options.fs access');
-        assert.ok(this.filename, 'Thrift must be constructed with a options.entryPoint');
-        /* eslint-enable max-len */
-        this.filename = path.resolve(this.filename);
-        source = this.fs.readFileSync(this.filename, 'ascii');
-        this.idls[this.filename] = source;
-    }
-
     this.exception = null;
 
-    // Separate compile/link passes permits forward references and cyclic
-    // references.
-    this.compile(source);
-    // We only link from the root Thrift object.
-    if (!options.noLink) {
-        this.link();
-    }
+    var model = this;
+    this._parse(this.filename, this.allowIncludeAlias, function (err) {
+        if (err) {
+            return model._exitOnError(loadedCb, err);
+        }
+
+        // Separate compile/link passes permits forward references and cyclic
+        // references.
+        try {
+            model.compile();
+            // We only link from the root Thrift object.
+            if (!options.noLink) {
+                model.link();
+            }
+        } catch (err) {
+            return model._exitOnError(loadedCb, err);
+        }
+        if (loadedCb) {
+            loadedCb(undefined, model);
+        }
+    });
 }
 
 Thrift.prototype.models = 'module';
 
 Thrift.prototype.Message = Message;
 
+Thrift.prototype._exitOnError = function _exitOnError(loadedCb, err) {
+    if (loadedCb) {
+        loadedCb(err, undefined);
+    } else {
+        throw err;
+    }
+}
+
+Thrift.prototype._syncFileLoader = function _syncFileLoader(fs) { 
+    return function(filename, cb) {
+        var source;
+        var error;
+        try {
+            /* eslint-disable max-len */
+            assert.ok(fs, filename + ': Thrift must be constructed with either a complete set of options.idls, options.asts, or options.fs access');
+            /* eslint-enable max-len */
+            source = fs.readFileSync(path.resolve(filename), 'ascii');
+        } catch (err) {
+            error = err;
+        }
+        cb(error, source);
+    };
+}
+
+Thrift.prototype._parse = function _parse(filename, allowIncludeAlias, cb) {
+    if (this.parsed[filename]) {
+        return cb(undefined);
+    }
+    this.parsed[filename] = true;
+
+    var model = this;
+    var loader = this.loader;
+    if (this.idls[filename] || this.asts[filename]) {
+        loader = function (_, cb) { cb(undefined, model.idls[filename]); };
+    }
```

Comment:
Does the guard at the top of this function protect us from needing this block? Good way to know is to see whether it is covered.